File: examples/widgets/calculator/calculator.cpp
#include <QtGui>

#include <math.h>

#include "button.h"
#include "calculator.h"

Calculator::Calculator(QWidget *parent)
    : QDialog(parent)
{
    sumInMemory = 0.0;
    sumSoFar = 0.0;
    factorSoFar = 0.0;
    waitingForOperand = true;

    display = new QLineEdit("0");
    display->setReadOnly(true);
    display->setAlignment(Qt::AlignRight);
    display->setMaxLength(15);
    display->installEventFilter(this);

    QFont font = display->font();
    font.setPointSize(font.pointSize() + 8);
    display->setFont(font);

    QColor digitColor(150, 205, 205);
    QColor backspaceColor(225, 185, 135);
    QColor memoryColor(100, 155, 155);
    QColor operatorColor(155, 175, 195);

    for (int i = 0; i < NumDigitButtons; ++i) {
	digitButtons[i] = createButton(QString::number(i), digitColor,
                                       SLOT(digitClicked()));
    }

    pointButton = createButton(tr("."), digitColor, SLOT(pointClicked()));
    changeSignButton = createButton(tr("±"), digitColor, SLOT(changeSignClicked()));

    backspaceButton = createButton(tr("Backspace"), backspaceColor,
                                   SLOT(backspaceClicked()));
    clearButton = createButton(tr("Clear"), backspaceColor, SLOT(clear()));
    clearAllButton = createButton(tr("Clear All"), backspaceColor.light(120),
                                  SLOT(clearAll()));

    clearMemoryButton = createButton(tr("MC"), memoryColor,
                                     SLOT(clearMemory()));
    readMemoryButton = createButton(tr("MR"), memoryColor, SLOT(readMemory()));
    setMemoryButton = createButton(tr("MS"), memoryColor, SLOT(setMemory()));
    addToMemoryButton = createButton(tr("M+"), memoryColor,
                                     SLOT(addToMemory()));

    divisionButton = createButton(tr("÷"), operatorColor,
                                  SLOT(multiplicativeOperatorClicked()));
    timesButton = createButton(tr("×"), operatorColor,
                               SLOT(multiplicativeOperatorClicked()));
    minusButton = createButton(tr("-"), operatorColor,
                               SLOT(additiveOperatorClicked()));
    plusButton = createButton(tr("+"), operatorColor,
                              SLOT(additiveOperatorClicked()));

    squareRootButton = createButton(tr("Sqrt"), operatorColor,
                                    SLOT(unaryOperatorClicked()));
    powerButton = createButton(tr("x²"), operatorColor,
                               SLOT(unaryOperatorClicked()));
    reciprocalButton = createButton(tr("1/x"), operatorColor,
                                    SLOT(unaryOperatorClicked()));
    equalButton = createButton(tr("="), operatorColor.light(120),
                               SLOT(equalClicked()));

    QGridLayout *mainLayout = new QGridLayout;
    mainLayout->setSizeConstraint(QLayout::SetFixedSize);

    mainLayout->addWidget(display, 0, 0, 1, 6);
    mainLayout->addWidget(backspaceButton, 1, 0, 1, 2);
    mainLayout->addWidget(clearButton, 1, 2, 1, 2);
    mainLayout->addWidget(clearAllButton, 1, 4, 1, 2);

    mainLayout->addWidget(clearMemoryButton, 2, 0);
    mainLayout->addWidget(readMemoryButton, 3, 0);
    mainLayout->addWidget(setMemoryButton, 4, 0);
    mainLayout->addWidget(addToMemoryButton, 5, 0);

    for (int i = 1; i < NumDigitButtons; ++i) {
        int row = ((9 - i) / 3) + 2;
        int column = ((i - 1) % 3) + 1;
        mainLayout->addWidget(digitButtons[i], row, column);
    }

    mainLayout->addWidget(digitButtons[0], 5, 1);
    mainLayout->addWidget(pointButton, 5, 2);
    mainLayout->addWidget(changeSignButton, 5, 3);

    mainLayout->addWidget(divisionButton, 2, 4);
    mainLayout->addWidget(timesButton, 3, 4);
    mainLayout->addWidget(minusButton, 4, 4);
    mainLayout->addWidget(plusButton, 5, 4);

    mainLayout->addWidget(squareRootButton, 2, 5);
    mainLayout->addWidget(powerButton, 3, 5);
    mainLayout->addWidget(reciprocalButton, 4, 5);
    mainLayout->addWidget(equalButton, 5, 5);
    setLayout(mainLayout);

    setWindowTitle(tr("Calculator"));
}

bool Calculator::eventFilter(QObject *target, QEvent *event)
{
    if (target == display) {
        if (event->type() == QEvent::MouseButtonPress
                || event->type() == QEvent::MouseButtonDblClick
                || event->type() == QEvent::MouseButtonRelease
                || event->type() == QEvent::ContextMenu) {
            QMouseEvent *mouseEvent = static_cast<QMouseEvent *>(event);
            if (mouseEvent->buttons() & Qt::LeftButton) {
                QPalette newPalette = palette();
                newPalette.setColor(QPalette::Base,
                                    display->palette().color(QPalette::Text));
                newPalette.setColor(QPalette::Text,
                                    display->palette().color(QPalette::Base));
                display->setPalette(newPalette);
            } else {
                display->setPalette(palette());
            }
            return true;
        }
    }
    return QDialog::eventFilter(target, event);
}

void Calculator::digitClicked()
{
    Button *clickedButton = qobject_cast<Button *>(sender());
    int digitValue = clickedButton->text().toInt();
    if (display->text() == "0" && digitValue == 0.0)
        return;

    if (waitingForOperand) {
        display->clear();
	waitingForOperand = false;
    }
    display->setText(display->text() + QString::number(digitValue));
}

void Calculator::unaryOperatorClicked()
{
    Button *clickedButton = qobject_cast<Button *>(sender());
    QString clickedOperator = clickedButton->text();
    double operand = display->text().toDouble();
    double result;

    if (clickedOperator == tr("Sqrt")) {
        if (operand < 0.0) {
            abortOperation();
            return;
        }
        result = sqrt(operand);
    } else if (clickedOperator == tr("x²")) {
        result = pow(operand, 2.0);
    } else if (clickedOperator == tr("1/x")) {
        if (operand == 0.0) {
	    abortOperation();
	    return;
        }
        result = 1.0 / operand;
    }
    display->setText(QString::number(result));
    waitingForOperand = true;
}

void Calculator::additiveOperatorClicked()
{
    Button *clickedButton = qobject_cast<Button *>(sender());
    QString clickedOperator = clickedButton->text();
    double operand = display->text().toDouble();

    if (!pendingMultiplicativeOperator.isEmpty()) {
        if (!calculate(operand, pendingMultiplicativeOperator)) {
            abortOperation();
	    return;
        }
        display->setText(QString::number(factorSoFar));
        operand = factorSoFar;
        factorSoFar = 0.0;
        pendingMultiplicativeOperator.clear();
    }

    if (!pendingAdditiveOperator.isEmpty()) {
        if (!calculate(operand, pendingAdditiveOperator)) {
            abortOperation();
	    return;
        }
        display->setText(QString::number(sumSoFar));
    } else {
        sumSoFar = operand;
    }

    pendingAdditiveOperator = clickedOperator;
    waitingForOperand = true;
}

void Calculator::multiplicativeOperatorClicked()
{
    Button *clickedButton = qobject_cast<Button *>(sender());
    QString clickedOperator = clickedButton->text();
    double operand = display->text().toDouble();

    if (!pendingMultiplicativeOperator.isEmpty()) {
        if (!calculate(operand, pendingMultiplicativeOperator)) {
            abortOperation();
	    return;
        }
        display->setText(QString::number(factorSoFar));
    } else {
        factorSoFar = operand;
    }

    pendingMultiplicativeOperator = clickedOperator;
    waitingForOperand = true;
}

void Calculator::equalClicked()
{
    double operand = display->text().toDouble();

    if (!pendingMultiplicativeOperator.isEmpty()) {
        if (!calculate(operand, pendingMultiplicativeOperator)) {
            abortOperation();
	    return;
        }
        operand = factorSoFar;
        factorSoFar = 0.0;
        pendingMultiplicativeOperator.clear();
    }
    if (!pendingAdditiveOperator.isEmpty()) {
        if (!calculate(operand, pendingAdditiveOperator)) {
            abortOperation();
	    return;
        }
        pendingAdditiveOperator.clear();
    } else {
        sumSoFar = operand;
    }

    display->setText(QString::number(sumSoFar));
    sumSoFar = 0.0;
    waitingForOperand = true;
}

void Calculator::pointClicked()
{
    if (waitingForOperand)
        display->setText("0");
    if (!display->text().contains("."))
        display->setText(display->text() + tr("."));
    waitingForOperand = false;
}

void Calculator::changeSignClicked()
{
    QString text = display->text();
    double value = text.toDouble();

    if (value > 0.0) {
        text.prepend(tr("-"));
    } else if (value < 0.0) {
        text.remove(0, 1);
    }
    display->setText(text);
}

void Calculator::backspaceClicked()
{
    if (waitingForOperand)
        return;

    QString text = display->text();
    text.chop(1);
    if (text.isEmpty()) {
        text = "0";
        waitingForOperand = true;
    }
    display->setText(text);
}

void Calculator::clear()
{
    if (waitingForOperand)
        return;

    display->setText("0");
    waitingForOperand = true;
}

void Calculator::clearAll()
{
    sumSoFar = 0.0;
    factorSoFar = 0.0;
    pendingAdditiveOperator.clear();
    pendingMultiplicativeOperator.clear();
    display->setText("0");
    waitingForOperand = true;
}

void Calculator::clearMemory()
{
    sumInMemory = 0.0;
}

void Calculator::readMemory()
{
    display->setText(QString::number(sumInMemory));
    waitingForOperand = true;
}

void Calculator::setMemory()
{
    equalClicked();
    sumInMemory = display->text().toDouble();
}

void Calculator::addToMemory()
{
    equalClicked();
    sumInMemory += display->text().toDouble();
}

Button *Calculator::createButton(const QString &text, const QColor &color,
                                 const char *member)
{
    Button *button = new Button(text, color);
    connect(button, SIGNAL(clicked()), this, member);
    return button;
}

void Calculator::abortOperation()
{
    clearAll();
    display->setText(tr("####"));
}

bool Calculator::calculate(double rightOperand, const QString &pendingOperator)
{
    if (pendingOperator == tr("+")) {
        sumSoFar += rightOperand;
    } else if (pendingOperator == tr("-")) {
        sumSoFar -= rightOperand;
    } else if (pendingOperator == tr("×")) {
        factorSoFar *= rightOperand;
    } else if (pendingOperator == tr("÷")) {
	if (rightOperand == 0.0)
	    return false;
	factorSoFar /= rightOperand;
    }
    return true;
}

File: examples/script/calculator/calculator.js
function Calculator(ui)
{
    this.ui = ui;

    this.pendingAdditiveOperator = "";
    this.pendingMultiplicativeOperator = "";
    this.sumInMemory = 0;
    this.sumSoFar = 0;
    this.factorSoFar = 0;
    this.waitingForOperand = true;

    this.abortOperation = function()
    {
        this.clearAll();
        this.ui.display.text = "####";
    }

    this.calculate = function(rightOperand, pendingOperator)
    {
        if (pendingOperator == "+") {
            this.sumSoFar += rightOperand;
        } else if (pendingOperator == "-") {
            this.sumSoFar -= rightOperand;
        } else if (pendingOperator == "*") {
            this.factorSoFar *= rightOperand;
        } else if (pendingOperator == "/") {
            if (rightOperand == 0)
            return false;
            this.factorSoFar /= rightOperand;
        }
        return true;
    }

    this.digitClicked = function()
    {
        var digitValue = this.sender.text - 0;
        if ((digitValue == 0) && (this.ui.display.text == "0"))
            return;
        if (this.waitingForOperand) {
            this.ui.display.clear();
            this.waitingForOperand = false;
        }
        this.ui.display.text += digitValue;
    }

    this.unaryOperatorClicked = function()
    {
        var operand = this.ui.display.text - 0;
        var result = 0;
        if (this.sender.text == "Sqrt") {
            if (operand < 0) {
                this.abortOperation();
                return;
            }
            result = Math.sqrt(operand);
        } else if (this.sender.text == "x^2") {
            result = Math.pow(operand, 2);
        } else if (this.sender.text == "1/x") {
            if (operand == 0.0) {
                this.abortOperation();
                return;
            }
            result = 1 / operand;
        }
        this.ui.display.text = result + "";
        this.waitingForOperand = true;
    }

    this.additiveOperatorClicked = function()
    {
        var operand = this.ui.display.text - 0;

        if (this.pendingMultiplicativeOperator.length != 0) {
            if (!this.calculate(operand, this.pendingMultiplicativeOperator)) {
                this.abortOperation();
            return;
            }
            this.ui.display.text = this.factorSoFar + "";
            operand = this.factorSoFar;
            this.factorSoFar = 0;
            this.pendingMultiplicativeOperator = "";
        }

        if (this.pendingAdditiveOperator.length != 0) {
            if (!this.calculate(operand, this.pendingAdditiveOperator)) {
                this.abortOperation();
                return;
            }
            this.ui.display.text = this.sumSoFar + "";
        } else {
            this.sumSoFar = operand;
        }

        this.pendingAdditiveOperator = this.sender.text;
        this.waitingForOperand = true;
    }

    this.multiplicativeOperatorClicked = function()
    {
        var operand = this.ui.display.text - 0;

        if (this.pendingMultiplicativeOperator.length != 0) {
            if (!this.calculate(operand, this.pendingMultiplicativeOperator)) {
                this.abortOperation();
            return;
            }
            this.ui.display.text = this.factorSoFar + "";
        } else {
            this.factorSoFar = operand;
        }

        this.pendingMultiplicativeOperator = this.sender.text;
        this.waitingForOperand = true;
    }

    this.equalClicked = function()
    {
        operand = this.ui.display.text - 0;

        if (this.pendingMultiplicativeOperator.length != 0) {
            if (!this.calculate(operand, this.pendingMultiplicativeOperator)) {
                this.abortOperation();
                return;
            }
            operand = this.factorSoFar;
            this.factorSoFar = 0.0;
            this.pendingMultiplicativeOperator = "";
        }
        if (this.pendingAdditiveOperator.length != 0) {
            if (!this.calculate(operand, this.pendingAdditiveOperator)) {
                this.abortOperation();
    	    return;
            }
            this.pendingAdditiveOperator = "";
        } else {
            this.sumSoFar = operand;
        }

        this.ui.display.text = this.sumSoFar + "";
        this.sumSoFar = 0.0;
        this.waitingForOperand = true;
    }

    this.pointClicked = function()
    {
        if (this.waitingForOperand)
            this.ui.display.text = "0";
        if (this.ui.display.text.indexOf(".") == -1)
            this.ui.display.text += ".";
        this.waitingForOperand = false;
    }

    this.changeSignClicked = function()
    {
        var text = this.ui.display.text;
        var value = text - 0;

        if (value > 0) {
            text = "-" + text;
        } else if (value < 0) {
            text = text.slice(1);
        }
        this.ui.display.text = text;
    }

    this.backspaceClicked = function()
    {
        if (this.waitingForOperand)
            return;

        var text = this.ui.display.text;
        text = text.slice(0, -1);
        if (text.length == 0) {
            text = "0";
            this.waitingForOperand = true;
        }
        this.ui.display.text = text;
    }

    this.clear = function()
    {
        if (this.waitingForOperand)
            return;

        this.ui.display.text = "0";
        this.waitingForOperand = true;
    }

    this.clearAll = function()
    {
        this.sumSoFar = 0.0;
        this.factorSoFar = 0.0;
        this.pendingAdditiveOperator = "";
        this.pendingMultiplicativeOperator = "";
        this.ui.display.text = "0";
        this.waitingForOperand = true;
    }

    this.clearMemory = function()
    {
        this.sumInMemory = 0.0;
    }

    this.readMemory = function()
    {
        this.ui.display.text = this.sumInMemory + "";
        this.waitingForOperand = true;
    }

    this.setMemory = function()
    {
        this.equalClicked();
        this.sumInMemory = this.ui.display.text - 0;
    }

    this.addToMemory = function()
    {
        this.equalClicked();
        this.sumInMemory += this.ui.display.text - 0;
    }

    with (ui) {
        display.text = "0";

        zeroButton.clicked.connect(this, this.digitClicked);
        oneButton.clicked.connect(this, "digitClicked");
        twoButton.clicked.connect(this, "digitClicked");
        threeButton.clicked.connect(this, "digitClicked");
        fourButton.clicked.connect(this, "digitClicked");
        fiveButton.clicked.connect(this, "digitClicked");
        sixButton.clicked.connect(this, "digitClicked");
        sevenButton.clicked.connect(this, "digitClicked");
        eightButton.clicked.connect(this, "digitClicked");
        nineButton.clicked.connect(this, "digitClicked");

        pointButton.clicked.connect(this, "pointClicked");
        changeSignButton.clicked.connect(this, "changeSignClicked");

        backspaceButton.clicked.connect(this, "backspaceClicked");
        clearButton.clicked.connect(this, "clear");
        clearAllButton.clicked.connect(this, "clearAll");

        clearMemoryButton.clicked.connect(this, "clearMemory");
        readMemoryButton.clicked.connect(this, "readMemory");
        setMemoryButton.clicked.connect(this, "setMemory");
        addToMemoryButton.clicked.connect(this, "addToMemory");
  
        divisionButton.clicked.connect(this, "multiplicativeOperatorClicked");
        timesButton.clicked.connect(this, "multiplicativeOperatorClicked");
        minusButton.clicked.connect(this, "additiveOperatorClicked");
        plusButton.clicked.connect(this, "additiveOperatorClicked");

        squareRootButton.clicked.connect(this, "unaryOperatorClicked");
        powerButton.clicked.connect(this, "unaryOperatorClicked");
        reciprocalButton.clicked.connect(this, "unaryOperatorClicked");
        equalButton.clicked.connect(this, "equalClicked");
    }
}